### PR TITLE
chore(ci): Speed up CI with filtering

### PR DESCRIPTION
**Pull request by NicholasLYang**

### CODE REVIEW

File: .github/workflows/pr-go-e2e-filtered.yml [hunk 1]
```diff
@@ -9,6 +9,9 @@ on:
     branches: ["main"]
   pull_request:
     types: [opened, edited, synchronize]
+    paths-ignore:
+      - "README.md"
```
Comment (mehulkar) [inline .github/workflows/pr-go-e2e-filtered.yml:13]: could be all `**.md*`

File: .github/workflows/pr-go-unit-filtered.yml [hunk 2]
```diff
@@ -9,6 +9,8 @@ on:
     branches: ["main"]
   pull_request:
     types: [opened, edited, synchronize]
+    paths:
+      - "**.go"
```
Comment (chris-olszewski) [inline .github/workflows/pr-go-unit-filtered.yml:13]: I think this needs to be expanded:
 - Go module files e.g. `go.mod` and `go.sum`
 - We have fixtures (`cli/internal/lockfile/testdata` and `cli/fixtures`) that should trigger unit tests if they change

**Comment by chris-olszewski** [comment]
It would be really nice to do some smoke tests for these rules. Unsure if it would work, but could you open some draft PR's targeting this branch and get some runs?

File: .github/workflows/check-codestyle-all.yml [hunk 3]
```diff
@@ -2,11 +2,12 @@ name: Check Code Style
 on:
   push:
     branches: [main, staging]
+    pull_request:
```
Comment (mehulkar) [inline .github/workflows/check-codestyle-all.yml:5]: is this intentional? did you want `on.pull_request` instead of `on.push.pull_request`?

**Review by mehulkar — COMMENTED** [review]
One nice side-effect of this will be that in addition to making select PRs green faster, it'll relieve some pressure off Github workflows. I'm guessing we're maxing out capacity / seeing longer queue times also?

File: .github/workflows/check-codestyle-all.yml (hunk 3 above)
Comment (NicholasLYang) [reply .github/workflows/check-codestyle-all.yml:5]: Ah yeah good catch

File: .github/workflows/pr-go-unit-filtered.yml (hunk 2 above)
Comment (NicholasLYang) [reply .github/workflows/pr-go-unit-filtered.yml:13]: Yeah I suppose we want to err on the side of too many paths versus too few. Should I just include the entire `cli` directory?

Comment (chris-olszewski) [reply .github/workflows/pr-go-unit-filtered.yml:13]: Probably? If we end up needing to get more specific we should probably err on the side of moving things out of `cli` than trying to enumerate everything that can affect the unit tests.

**Comment by NicholasLYang** [comment]
@chris-olszewski tested in  #2583. Since we're pretty conservative on filtering, if you change anything other than docs or markdown you'll trigger almost all of the workflows. We can make this stricter later on if necessary.

File: .github/workflows/pr-go-e2e-all.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+name: CLI E2E tests
```
Comment (mehulkar) [inline .github/workflows/pr-go-e2e-all.yml:1]: does the name need to be unique?

File: .github/workflows/pr-go-e2e-all.yml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+jobs:
+  test:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+    steps:
+      - run: 'echo "No test run required"'
```
Comment (mehulkar) [inline .github/workflows/pr-go-e2e-all.yml:16]: Why do we need this in addition to the filtered one?

**Review by mehulkar — APPROVED** [review]

**Review by chris-olszewski — APPROVED** [review]

File: .github/workflows/pr-go-e2e-all.yml (hunk 4 above)
Comment (NicholasLYang) [reply .github/workflows/pr-go-e2e-all.yml:1]: Actually it needs to be the same as the filtered

File: .github/workflows/pr-go-e2e-all.yml (hunk 5 above)
Comment (NicholasLYang) [reply .github/workflows/pr-go-e2e-all.yml:16]: For PR checks to pass. Annoying GitHub workaround lol

**Pull request merged**